### PR TITLE
Verify bucket type existence for list keys and buckets operations

### PR DESCRIPTION
**Pull request by jburwell**
Per defect #875, the client hangs when attempting to list keys or buckets for an undefined bucket types.

This fix represents a time/risk compromise for the [previous proposed solution](https://github.com/basho/riak_kv/pull/958) which made a deeper change to the riak_core_coverage_fsm -- pushing the check for bucket definition down to the state machine ensuring that all bucket type uses are consistently checked.  #965 has been opened to revisit this approach when we have the time to address the issue more holistically.

Supersedes previous FSM-based implementation [PR 958](https://github.com/basho/riak_kv/pull/958).

/cc @kellymclaughlin @seancribbs @reiddraper

### CODE REVIEW

**Comment by reiddraper** [comment]
Should this be on the 2.0-RC milestone? It's on '2.0' right now.

**Comment by jburwell** [comment]
@reiddraper mis-click -- fixed.

**Comment by reiddraper** [comment]
no prob, thanks

File: src/riak_kv_pb_bucket.erl [hunk 1]
```diff
@@ -95,17 +95,22 @@ encode(Message) ->
...
+		   {ok, ReqId} = C:stream_list_buckets(none, T, Type),
+		   {reply, {stream, ReqId}, State#state{req = Req, req_ctx = ReqId}};
+	       _ ->
+		   case C:list_buckets(none, T, Type) of
+		       {ok, Buckets} ->
+			   {reply, #rpblistbucketsresp{buckets = Buckets}, State};
+		       {error, Reason} ->
+			   {error, {format, Reason}, State}
+		   end
+	   end;
```
Comment (reiddraper) [inline src/riak_kv_pb_bucket.erl:108]: could this be made a bit more succinct by casing on both `maybe_bucket_type_exists` and `S`, at once:

``` erlang
case {maybe_bucket_type_exists(Type), S} of
    {true, true} ->
        ...;
    _ ->
        ...
end
```

File: src/riak_kv_pb_bucket.erl [hunk 2]
```diff
@@ -161,11 +170,22 @@ process_stream({ReqId, Error}, ReqId,
                State=#state{ req=#rpblistbucketsreq{}, req_ctx=ReqId}) ->
     {error, {format, Error}, State#state{req = undefined, req_ctx = undefined}}.
 
+maybe_bucket_type_exists(undefined) ->
```
Comment (reiddraper) [inline src/riak_kv_pb_bucket.erl:171]: It would be great to have a spec here.

Comment (kellymclaughlin) [inline src/riak_kv_pb_bucket.erl:171]: +1 to adding specs for all the `maybe` helpers

**Comment by jburwell** [comment]
@reiddraper Updated with specs and simplified case statement.  Unfortunately, git would not squash the commits without detaching the HEAD and preventing a clean push.

**Comment by reiddraper** [comment]
@jburwell you wanna ping me on hipchat and I can help with some of the git issues you've been running into?

**Comment by kellymclaughlin** [comment]
- [X] eunit + eqc
- [X] dialyzer
- [X] xref
- [X] riak_test

**Comment by kellymclaughlin** [comment]
The change looks good to me now. How about we squash the commits on a new branch and push that up for merge? The convention we usually use is `<current-branch-name>-rebased`. Alternatively you can squash and force push the branch as long as you have run:

```
git config --global push.default current
```

**Comment by jburwell** [comment]
Superseded by [PR 970](https://github.com/basho/riak_kv/pull/970) to properly squash commits and clean up git history.

**Pull request closed without merging**